torch/csrc/DataLoader.cpp

```diff
@@ -0,0 +1,191 @@
+  sa.sa_flags = SA_RESTART;                                                   \
+  sigemptyset(&sa.sa_mask);                                                   \
+  ERROR_FLAG |= sigaction(SIGNAL, &sa, NULL) != 0;                            \
+}
+
+SIGNAL_HANDLER(SIGBUS, handler_SIGBUS, "ERROR: Unexpected bus error encountered in worker. "
+  "This might be caused by insufficient shared memory (shm).\n");
+SIGNAL_HANDLER(SIGSEGV, handler_SIGSEGV, "ERROR: Unexpected segmentation fault encountered in worker.\n");
+
+static std::vector<pid_t> worker_pid_vec = {};
+// The following are needed since std::vector is not asynchronous safe.
+static std::atomic<pid_t *> worker_pids;
+static std::atomic<size_t> num_worker_pids(0);
+static std::atomic<size_t> num_dataloaders(0);
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
+  HANDLE_TH_ERRORS
+  int error = 0;
+  SET_SIGNAL_HANDLER(SIGBUS, &handler_SIGBUS, error);
+  SET_SIGNAL_HANDLER(SIGSEGV, &handler_SIGSEGV, error);
+  if (error == 0) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
+  END_HANDLE_TH_ERRORS
+}
+
+// SIGCHLD hander should be registered on main loader process to catch any
+// worker failing. SIGALRM handler is needed for implementing timeout.
+// Python handles are _set_main_signal_handers() and
+// _remove_main_signal_handers().
+static void handler_SIGCHLD_main(int sig) {
+  int status;
+  pid_t p;
+  pid_t *pid_ptr = worker_pids;
+
+  // Only check the pids we care about so that Python can see other processes'
+  // status.
+  for (size_t i = 0; i < num_worker_pids; i++) {
+    // The flags and status checks ensure that we are really observing a child
+    // exiting, rather than other cases such as SIGSTOP and SIGCONT.
+    // https://stackoverflow.com/a/40707100
+    p = waitpid(*pid_ptr, &status, WNOHANG|WUNTRACED|WCONTINUED);
+    // Ignore errors / no changes
+    if (p <= 0)
+      continue;
+    if (WIFCONTINUED(status) || WIFSTOPPED(status))
+      continue;
+    if (WIFEXITED(status) != 0 && WEXITSTATUS(status) == 0)
+      continue;
+    _exit(EXIT_FAILURE);
+    pid_ptr++;
+  }
+}
+
+// We don't want to exit on any SIGCHLD from any child. child_pids is a sequence
+// of pids we are interested in.
+PyObject *THPModule_setMainSignalHandlers(PyObject *module, PyObject *child_pids) {
+  HANDLE_TH_ERRORS
+  // assert these types are lock free, just to be safe
+  THPUtils_assert(worker_pids.is_lock_free(), "worker_pids is not lock free");
+  THPUtils_assert(num_worker_pids.is_lock_free(), "num_worker_pids is not lock free");
+
+  THPUtils_assert(PyTuple_Check(child_pids), "_set_main_signal_handler "
+        "expects a list, but got %s", THPUtils_typename(child_pids));
+
+  num_dataloaders++;
+  auto size = PyTuple_GET_SIZE(child_pids);
+  for (int idx = 0; idx < size; idx++) {
+    PyObject* obj = PyTuple_GET_ITEM(child_pids, idx);
+    worker_pid_vec.push_back((pid_t) THPUtils_unpackLong(obj));
```

Comment:
I removed the lock here since, if I understand correctly, GIL is always held unless I give it up. Please correct me if I am wrong :)